homeassistant/components/matter/update.py

```diff
@@ -0,0 +1,221 @@
+            == clusters.OtaSoftwareUpdateRequestor.Enums.UpdateStateEnum.kDownloading
+            and update_progress is not None
+            and update_progress > 0
+        ):
+            self._attr_in_progress = update_progress
+        else:
+            self._attr_in_progress = True
+
+    async def async_update(self) -> None:
+        """Call when the entity needs to be updated."""
+        try:
+            update_information = await self.matter_client.check_node_update(
+                node_id=self._endpoint.node.node_id
+            )
+            if not update_information:
+                self._attr_latest_version = self._attr_installed_version
+                return
+
+            self._software_update = update_information
+            self._attr_latest_version = update_information.software_version_string
+            self._attr_release_url = update_information.release_notes_url
+        except UpdateCheckError as err:
+            raise HomeAssistantError(f"Error finding applicable update: {err}") from err
+
+    async def async_added_to_hass(self) -> None:
+        """Call when the entity is added to hass."""
+        await super().async_added_to_hass()
+
+        if state := await self.async_get_last_state():
+            self._attr_latest_version = state.attributes.get(ATTR_LATEST_VERSION)
+
+        if (extra_data := await self.async_get_last_extra_data()) and (
+            matter_extra_data := MatterUpdateExtraStoredData.from_dict(
+                extra_data.as_dict()
+            )
+        ):
+            self._software_update = matter_extra_data.software_update
+        else:
+            # Check for updates when added the first time.
+            await self.async_update()
+
+    @property
+    def extra_restore_state_data(self) -> MatterUpdateExtraStoredData:
+        """Return Matter specific state data to be restored."""
+        return MatterUpdateExtraStoredData(self._software_update)
+
+    @property
+    def entity_picture(self) -> str | None:
+        """Return the entity picture to use in the frontend."""
+
+        # The Matter brand picture is not appropriate as this is the update
+        # entity for the device.
+        return None
+
+    async def async_install(
+        self, version: str | None, backup: bool, **kwargs: Any
+    ) -> None:
+        """Install a new software version."""
+
+        software_version: str | int | None = version
+        if self._software_update is not None and (
+            version is None or version == self._software_update.software_version_string
+        ):
+            # Update to the version previously fetched and shown.
+            # We can pass the integer version directly to speedup download.
+            software_version = self._software_update.software_version
+
+        if software_version is None:
+            raise HomeAssistantError("No software version specified")
+
+        self._attr_in_progress = True
+        self.async_write_ha_state()
```

Comment:
Why is this needed? If the entity is polled, an update should be scheduled after service calls

Reply:
It usually takes 2-3s for the device to get the first update progress state changes. Pushing the progress state change here, really improves user expirience: With that, the frontend immeaditly switches into a "in progress" animation. Without it, the frontend doesn't update, and it feels like the "Install" button is broken.

Reply:
It usually takes 2-3s for the device to get the first update progress state changes. Pushing the progress state change here, really improves user expirience: With that, the frontend immeaditly switches into a "in progress" animation. Without it, the frontend doesn't update, and it feels like the "Install" button is broken.

Reply:
OK, please add a comment explaining we immediately update the state to not have to wait for the real update state change.